client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoPreparedStatement.java

```diff
@@ -109,14 +109,29 @@ public class TrinoPreparedStatement
     private final String statementName;
     private final String originalSql;
     private boolean isBatch;
+    private boolean prepareStatementExecuted;
 
     TrinoPreparedStatement(TrinoConnection connection, Consumer<TrinoStatement> onClose, String statementName, String sql)
             throws SQLException
     {
         super(connection, onClose);
         this.statementName = requireNonNull(statementName, "statementName is null");
         this.originalSql = requireNonNull(sql, "sql is null");
-        super.execute(format("PREPARE %s FROM %s", statementName, sql));
+        if (connection().isUseLegacyPreparedStatements()) {
+            super.execute(format("PREPARE %s FROM %s", statementName, sql));
+            prepareStatementExecuted = true;
+        }
+    }
+
+    private void prepareStatementIfNecessary()
```

Comment:
why is this needed? it's not obvious to me? Won't this always be executed in the constructor?

Reply:
Also nit: move helpers AFTER the methods that use them

Reply:
This method is to avoid the `PREPARE` statement to be called multiple times if `getMetadata()` or `getParameterMetadata()` are called multiple times.

The constructor will only execute the `PREPARE` statement if `connection().isUseLegacyPreparedStatements()` is true.

Reply:
Moved Move the method and added a comment